### ISSUE TITLE
RUSTC_WRAPPER is applied even after being unset

### ISSUE THREAD

**Issue by jplatte** [rust-lang/cargo#9553]
**Problem**

Recently, I ran `cargo bloat --time` on one a Cargo project. Afterwards, I noticed there are now `json-time {...}` output lines when running other `cargo` commands (`check` for example). It looks like these are how cargo-bloat collects the compilation times (https://github.com/RazrFalcon/cargo-bloat/blob/ddbb5f55088eb7e28c0cc80e7a1fdfb1310c19b5/src/main.rs#L399), through a `RUSTC_WRAPPER` executable.

I have verified that at the time I run `cargo check`, no `RUSTC_WRAPPER` is set so I'm assuming it's being cached by Cargo somewhere.

**Steps**

1. Install [cargo-bloat](https://github.com/RazrFalcon/cargo-bloat)
2. Run `cargo bloat --time` on some Cargo project
3. Run `cargo check`

**Notes**

Tested with both `cargo 1.53.0-beta (4369396ce 2021-04-27)` and `cargo 1.54.0-nightly (0cecbd673 2021-06-01)`.

**Comment by gbaranski** [comment]
Same issue, solved with `cargo clean` in project where `cargo bloat --time` has been used.

Not sure if uninstalling `cargo-bloat` and removing anything related to it from `~/.cargo` helped anything, but that might be worth mentioning in caes `cargo clean` didn't help.

**Comment by weihanglo** [comment]
Cargo indeed caches the stderr of rustc invocation and then replays if it's a fresh re-compilation.

https://github.com/rust-lang/cargo/blob/df8cda0a12108a4b2959af6b1cae2ced47bb0b64/src/cargo/core/compiler/mod.rs#L178-L186

In addition, lines do not look like JSON at start are also cached:

https://github.com/rust-lang/cargo/blob/df8cda0a12108a4b2959af6b1cae2ced47bb0b64/src/cargo/core/compiler/mod.rs#L1302-L1305

In general, rustc should emit stderr in JSON, but I doubt that we can blindly disable caching on those not starting with a `{`.

**Comment by jplatte** [comment]
> Cargo indeed caches the stderr of rustc invocation and then replays if it's a fresh re-compilation.

Shouldn't `RUSTC_WRAPPER` changing invalidate all cached rustc invocation output though?

**Comment by weihanglo** [comment]
It seems that [fingerprint calculation] does not include `RUSTC_WRAPPER` change. I don't know if there is a reason, but there are cases that cargo-fix and clippy leverage the other wrapper env `RUSTC_WORKSPACE_WRAPPER`. Invalidating all rustc invocation for that may lead to full recompilation.

[fingerprint calculation]:  https://github.com/rust-lang/cargo/blob/e1fb17631eb1b3665cdbe45b1c186111577ef512/src/cargo/core/compiler/fingerprint.rs#L1215-L1228

**Comment by jplatte** [comment]
It seems like `RUSTC_WRAPPER`s should be able to decide whether they get the same cache as other invocations...

Alternatively, maybe `cargo-bloat --time` could be updated to set a different cache directory (and delete it after it is done), instead of cleaning the regular target dir as I think it does now. Would that solve this issue? cc @RazrFalcon

**Comment by RazrFalcon** [comment]
Yes, `cargo-bloat` has to be updated. I just didn't had time to do it.

**Comment by jplatte** [comment]
Alright, so is `RUSTC_WRAPPER` emitting extra (non-JSON) output just not meant to happen and this issue can be closed?

**Comment by RazrFalcon** [comment]
I still think this is a Cargo issue.
